core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -717,18 +717,20 @@ class KafkaApis(val requestChannel: RequestChannel,
   }
 
   private[server] def fetchOffsetsBefore(log: Log, timestamp: Long, maxNumOffsets: Int): Seq[Long] = {
-    val segsArray = log.logSegments.toArray
-    var offsetTimeArray: Array[(Long, Long)] = null
-    val lastSegmentHasSize = segsArray.last.size > 0
-    if (lastSegmentHasSize)
-      offsetTimeArray = new Array[(Long, Long)](segsArray.length + 1)
-    else
-      offsetTimeArray = new Array[(Long, Long)](segsArray.length)
+    // cache to avoid race conditions
+    val segments = log.logSegments.toSeq
+    val lastSegmentHasSize = segments.last.size > 0
+
+    val offsetTimeArray =
+      if (lastSegmentHasSize)
+        new Array[(Long, Long)](segments.length + 1)
+      else
+        new Array[(Long, Long)](segments.length)
 
-    for (i <- segsArray.indices)
-      offsetTimeArray(i) = (segsArray(i).baseOffset, segsArray(i).lastModified)
+    for ((segment, i) <- segments.zipWithIndex)
+      offsetTimeArray(i) = (segment.baseOffset, segment.lastModified)
```

Comment:
Indeed, with the use of `Buffer` this can be reverted. I'm going to take a look at the tests and commit them together if I'll manage to add something.